### PR TITLE
sessionctx/stmtctx: do not use copy-on-read for GetWarnings function

### LINKED ISSUE #39702: Extremely inefficient GetWarnings() makes the query slow
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

Originally encountered by an user https://asktug.com/t/topic/994126
A minimal reproduce can be:

```
CREATE TABLE `t2` (`col` decimal(20,4) DEFAULT NULL);
INSERT INTO `t2` VALUES (round(rand() * 0.49 + 0.0123, 4));
....  (repeat several times)
INSERT INTO `t2` SELECT * FROM `t2`;
.... (repeat several times to construct enough data)

CREATE TABLE `t1` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `c1` decimal(20,2) DEFAULT NULL,
  `create_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `update_time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
)
INSERT INTO t1 (c1) SELECT * FROM t2;
```


### 2. What did you expect to see? (Required)

The last query finish in maybe 5~10s

### 3. What did you see instead (Required)

It takes nearly 8min to run the query!

```
mysql> insert into t1 (c1) select * from t2;
Query OK, 491520 rows affected, 65535 warnings (7 min 53.87 sec)
Records: 491520  Duplicates: 0  Warnings: 65535
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

nightly

```
mysql> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                           |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.5.0-alpha
Edition: Community
Git Commit Hash: dc6d9a01112a54af1ac7d0200521ea20b6d09b30
Git Branch: heads/refs/tags/v6.5.0-alpha
UTC Build Time: 2022-12-06 14:32:10
GoVersion: go1.19.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #39702

Problem Summary:

### What is changed and how it works?

7 min 53s => 9.32s
(B.T.W, if we don't add error stack on decimal truncate error, the execution time can furthur reduce to 5s)

image

Before the change:

```
mysql> insert into t1 (c1) select * from t2;
Query OK, 491520 rows affected, 65535 warnings (7 min 53.87 sec)
Records: 491520  Duplicates: 0  Warnings: 65535
```

After:

```
mysql> insert into t1 (c1) select * from t2;
Query OK, 491520 rows affected, 65535 warnings (9.32 sec)
Records: 491520  Duplicates: 0  Warnings: 65535
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)

See https://github.com/pingcap/tidb/issues/39702 reproduce process.

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -730,9 +730,7 @@ func (sc *StatementContext) SetMessage(msg string) {
 func (sc *StatementContext) GetWarnings() []SQLWarn {
 	sc.mu.Lock()
 	defer sc.mu.Unlock()
-	warns := make([]SQLWarn, len(sc.mu.warnings))
-	copy(warns, sc.mu.warnings)
-	return warns
+	return sc.mu.warnings
```
Comment (Defined2014) [inline sessionctx/stmtctx/stmtctx.go:733]: Will return warnings directly cause conflict?

Comment (tiancaiamao) [reply sessionctx/stmtctx/stmtctx.go:732]: This code date back to very early commit:

```
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 407) // GetWarnings gets warnings.
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 408) func (sc *StatementContext) GetWarnings() []error {
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 409)        sc.mu.Lock()
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 410)        warns := make([]error, len(sc.mu.warnings))
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 411)        copy(warns, sc.mu.warnings)
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 412)        sc.mu.Unlock()
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 413)        return warns
5230f5eabd7 (Ewan Chou    2016-11-29 14:25:38 +0800 414) }
```

**The commit is sooooooo old that it's hard to figure out why it's written like that.**

The change should work, if not, at least we can see where the data race happen and try to fix again.
@Defined2014

Comment (Defined2014) [reply sessionctx/stmtctx/stmtctx.go:732]: without copy, does the lock necessary here?

Comment (tiancaiamao) [reply sessionctx/stmtctx/stmtctx.go:732]: > without copy, does the lock necessary here?

I think so ... `GetWarnings / SetWarnings` may be visited by many different component, maybe concurrently.

Comment (Defined2014) [reply sessionctx/stmtctx/stmtctx.go:732]: Got it. The value of `sc.mu.warnings` will be changed by SetWarnings.

**Review by Defined2014 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**